### PR TITLE
fix(cairo): fix reader reload after truncate table

### PR DESCRIPTION
**Pull request by ideoma**
When table is truncated, `TableReader` did not truck this in the `reload()` method and only reload partitions if they were missing. With the fix TableReader checks `truncateVersion` in `_txn` file and when changed all fully reloads all the partitions.
Attach and drop partitions forces same full reload on `TableReader`

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 89 / 120 (74.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|49|79|62.03%|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 176 / 178 (98.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|85|86|98.84%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 176 / 178 (98.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|85|86|98.84%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 176 / 178 (98.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|85|86|98.84%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|2|2|100.00%|

File: core/src/main/java/io/questdb/cairo/SymbolMapWriter.java [hunk 1]
```diff
@@ -212,6 +212,8 @@ public void truncate() {
         final int symbolCapacity = offsetMem.getInt(HEADER_CAPACITY);
         offsetMem.truncate();
         offsetMem.putInt(HEADER_CAPACITY, symbolCapacity);
+        offsetMem.putBool(HEADER_CACHE_ENABLED, isCached());
+        updateNullFlag(nullValue = false);
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/SymbolMapWriter.java:216]: nit: two separate lines would read better.

File: core/src/test/java/io/questdb/cairo/FullFwdDataFrameCursorTest.java [hunk 2]
```diff
@@ -2005,7 +2005,7 @@ private void testReplaceIndexedColWithIndexedWithTruncate(int partitionBy, long
                     cursor.of(reader);
                     record.of(reader);
 
-                    assertSymbolFoundInIndex(cursor, record, 1, M);
+                    assertSymbolFoundInIndex(cursor, record, 1, M); // VTJWCPSW
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/cairo/FullFwdDataFrameCursorTest.java:2008]: nit: what does "VTJWCPSW" mean?

File: core/src/main/java/io/questdb/cairo/TxReader.java [hunk 3]
```diff
@@ -209,6 +214,7 @@ public boolean unsafeLoadAll() {
             final long prevPartitionTableVersion = this.partitionTableVersion;
             this.partitionTableVersion = getLong(TableUtils.TX_OFFSET_PARTITION_TABLE_VERSION_64);
             this.columnVersion = unsafeReadColumnVersion();
+            this.truncateVersion = getLong(TableUtils.TX_OFFSET_TRUNCATE_VERSION_64);
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/TxReader.java:217]: We need to update `dumpTo()` method in this class as well to make sure that the truncate version is included into snapshots. Also it'd be nice to update _txn file's contents verification in `SnapshotTest`.

File: core/src/main/java/io/questdb/cairo/SymbolMapWriter.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/SymbolMapWriter.java]: done

File: core/src/test/java/io/questdb/cairo/FullFwdDataFrameCursorTest.java (hunk 2 above)
Comment (ideoma) [reply core/src/test/java/io/questdb/cairo/FullFwdDataFrameCursorTest.java]: debris, removed

File: core/src/main/java/io/questdb/cairo/TxReader.java (hunk 3 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TxReader.java:217]: done

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 178 / 180 (98.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|85|86|98.84%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|3|3|100.00%|

**Review by puzpuzpuz — DISMISSED** [review]
LGTM. Left a few questions.

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 4]
```diff
@@ -590,12 +590,6 @@ public int attachPartition(long timestamp) {
             return PARTITION_ALREADY_ATTACHED;
         }
 
-        if (metadata.getSymbolMapCount() > 0) {
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java:593]: You right, I'll update docs

File: core/src/test/java/io/questdb/griffin/AlterTableAttachPartitionTest.java [hunk 5]
```diff
@@ -35,7 +35,6 @@
 import io.questdb.std.str.Path;
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/test/java/io/questdb/griffin/AlterTableAttachPartitionTest.java:35]: Added tests

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 6]
```diff
@@ -1554,15 +1554,77 @@ private static ColumnVersionWriter openColumnVersionFile(FilesFacade ff, Path pa
...
+                    case ColumnType.BOOLEAN:
+                    case ColumnType.BYTE:
+                    case ColumnType.TIMESTAMP:
+                    case ColumnType.DATE:
+                    case ColumnType.DOUBLE:
+                    case ColumnType.CHAR:
+                    case ColumnType.SHORT:
+                    case ColumnType.FLOAT:
+                    case ColumnType.LONG256:
+                        // Consider Symbols as fixed, check data file size
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:1598]: nit: why do we have this comment here? This line and what's above seems to be unrelated with symbols.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 205 / 208 (98.56%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|110|112|98.21%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|3|3|100.00%|

**Review by puzpuzpuz — DISMISSED** [review]
Thanks for adding validation for index files!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 205 / 208 (98.56%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|110|112|98.21%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|3|3|100.00%|

**Comment by puzpuzpuz** [comment]
I may have overlooked one more place that we might need to update: `io.questdb.cliutil.TxSerializer` and `TxFileStruct` classes.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 205 / 208 (98.56%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|110|112|98.21%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 205 / 208 (98.56%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnVersionWriter.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|110|112|98.21%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|73|73|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/Mig620.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|3|3|100.00%|

**Review by puzpuzpuz — APPROVED** [review]
